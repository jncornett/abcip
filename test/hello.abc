# abcip is a simple tool for crafting packets
# that go from point a to point b.  this file
# is a an example that shows how to do the
# basics.

# by now you know that comments start with '#'
# and go to end of line.

# there are four types of statements: a, b, c,
# and d.  d statements define things that apply
# to both ends.

# the default stack is eth:ip4:tcp
# but we define it here to show how
d ( stack="eth:ip4:tcp" )
#   layer:  2   3   4

# c statements configure things for each end
# specifically.
c ( 
   3:a="10.4.10.182", # host a ip4 address
   4:b=80             # host b port
)

# we know 3:a is the ip4 address because our
# stack has ip4 at layer 3 and since layer 4
# was defined to be tcp, 4:b is a port.

# a and b statements generate packets from a to
# b and from b to a respectively.

# this is a normal 3-way handshake
# from client a to server b:
a ( syn ) 
b ( syn, ack )
a ( ack )

# by default, attributes apply to the top layer
# so we didn't need to say a ( 4:syn ).

# now a sends some data
a ( ack, pay="hello world" )
b ( ack ) # and b acks it

# sequence and acknowledgement numbers are
# calculated automatically for normal traffic

# b replies with 11 bytes of default data
b( ack, len=11 )

# a acks it and closes the connection
a ( fin, ack )
b ( fin, ack )
a ( ack )

# that's it.  you can create a full tcp session
# pcap with abcip --pcap hello.pcap < hello.abc.

# to generate ip6 instead of ip4 add the option:
# --stack "eth:ip6:tcp".

