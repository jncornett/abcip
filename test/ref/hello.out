IP (tos 0x0, ttl 64, id 1, offset 0, flags [none], proto TCP (6), length 40)
    10.4.10.182.48620 > 10.9.8.7.80: Flags [S], cksum 0xaadb (correct), seq 1, win 8192, length 0
	0x0000:  0209 0807 0605 0201 0203 0405 0800 4500  ..............E.
	0x0010:  0028 0001 0000 4006 5406 0a04 0ab6 0a09  .(....@.T.......
	0x0020:  0807 bdec 0050 0000 0001 0000 0000 5002  .....P........P.
	0x0030:  2000 aadb 0000                           ......
IP (tos 0x0, ttl 64, id 1, offset 0, flags [none], proto TCP (6), length 40)
    10.9.8.7.80 > 10.4.10.182.48620: Flags [S.], cksum 0xaac9 (correct), seq 1, ack 2, win 8192, length 0
	0x0000:  0201 0203 0405 0209 0807 0605 0800 4500  ..............E.
	0x0010:  0028 0001 0000 4006 5406 0a09 0807 0a04  .(....@.T.......
	0x0020:  0ab6 0050 bdec 0000 0001 0000 0002 5012  ...P..........P.
	0x0030:  2000 aac9 0000                           ......
IP (tos 0x0, ttl 64, id 2, offset 0, flags [none], proto TCP (6), length 40)
    10.4.10.182.48620 > 10.9.8.7.80: Flags [.], cksum 0xaaca (correct), ack 1, win 8192, length 0
	0x0000:  0209 0807 0605 0201 0203 0405 0800 4500  ..............E.
	0x0010:  0028 0002 0000 4006 5405 0a04 0ab6 0a09  .(....@.T.......
	0x0020:  0807 bdec 0050 0000 0002 0000 0002 5010  .....P........P.
	0x0030:  2000 aaca 0000                           ......
IP (tos 0x0, ttl 64, id 3, offset 0, flags [none], proto TCP (6), length 51)
    10.4.10.182.48620 > 10.9.8.7.80: Flags [.], cksum 0x18f1 (correct), seq 1:12, ack 1, win 8192, length 11
	0x0000:  0209 0807 0605 0201 0203 0405 0800 4500  ..............E.
	0x0010:  0033 0003 0000 4006 53f9 0a04 0ab6 0a09  .3....@.S.......
	0x0020:  0807 bdec 0050 0000 0002 0000 0002 5010  .....P........P.
	0x0030:  2000 18f1 0000 6865 6c6c 6f20 776f 726c  ......hello.worl
	0x0040:  64                                       d
IP (tos 0x0, ttl 64, id 2, offset 0, flags [none], proto TCP (6), length 40)
    10.9.8.7.80 > 10.4.10.182.48620: Flags [.], cksum 0xaabf (correct), ack 12, win 8192, length 0
	0x0000:  0201 0203 0405 0209 0807 0605 0800 4500  ..............E.
	0x0010:  0028 0002 0000 4006 5405 0a09 0807 0a04  .(....@.T.......
	0x0020:  0ab6 0050 bdec 0000 0002 0000 000d 5010  ...P..........P.
	0x0030:  2000 aabf 0000                           ......
IP (tos 0x0, ttl 64, id 3, offset 0, flags [none], proto TCP (6), length 51)
    10.9.8.7.80 > 10.4.10.182.48620: Flags [.], cksum 0x34d9 (correct), seq 1:12, ack 12, win 8192, length 11
	0x0000:  0201 0203 0405 0209 0807 0605 0800 4500  ..............E.
	0x0010:  0033 0003 0000 4006 53f9 0a09 0807 0a04  .3....@.S.......
	0x0020:  0ab6 0050 bdec 0000 0002 0000 000d 5010  ...P..........P.
	0x0030:  2000 34d9 0000 4c6f 7265 6d20 6970 7375  ..4...Lorem.ipsu
	0x0040:  6d                                       m
IP (tos 0x0, ttl 64, id 4, offset 0, flags [none], proto TCP (6), length 40)
    10.4.10.182.48620 > 10.9.8.7.80: Flags [F.], cksum 0xaab3 (correct), seq 12, ack 12, win 8192, length 0
	0x0000:  0209 0807 0605 0201 0203 0405 0800 4500  ..............E.
	0x0010:  0028 0004 0000 4006 5403 0a04 0ab6 0a09  .(....@.T.......
	0x0020:  0807 bdec 0050 0000 000d 0000 000d 5011  .....P........P.
	0x0030:  2000 aab3 0000                           ......
IP (tos 0x0, ttl 64, id 4, offset 0, flags [none], proto TCP (6), length 40)
    10.9.8.7.80 > 10.4.10.182.48620: Flags [F.], cksum 0xaab2 (correct), seq 12, ack 13, win 8192, length 0
	0x0000:  0201 0203 0405 0209 0807 0605 0800 4500  ..............E.
	0x0010:  0028 0004 0000 4006 5403 0a09 0807 0a04  .(....@.T.......
	0x0020:  0ab6 0050 bdec 0000 000d 0000 000e 5011  ...P..........P.
	0x0030:  2000 aab2 0000                           ......
IP (tos 0x0, ttl 64, id 5, offset 0, flags [none], proto TCP (6), length 40)
    10.4.10.182.48620 > 10.9.8.7.80: Flags [.], cksum 0xaab2 (correct), ack 13, win 8192, length 0
	0x0000:  0209 0807 0605 0201 0203 0405 0800 4500  ..............E.
	0x0010:  0028 0005 0000 4006 5402 0a04 0ab6 0a09  .(....@.T.......
	0x0020:  0807 bdec 0050 0000 000e 0000 000e 5010  .....P........P.
	0x0030:  2000 aab2 0000                           ......
